### PR TITLE
Fix fair follows

### PR DESCRIPTION
**Pull request by mbilokonsky**
This adds a new field to fairs in metaphysics, `followed_content`. Inside of followed content are `artists` and `galleries`, which embed raw arrays containing artwork and galleries that the current user follows AND which are affiliated with this fair.

This is tricky to test, and gravity caching seems to make it a little trickier. The bulk of the work here was actually around getting viable data into gravity that represented sufficient complexity to test that this code works as expected. Right now, I just ran the data setup code you'll find below, so this data all lives on the gravity staging server. (It'll be wiped over the weekend, though, so someone should rerun it on Monday if necessary!)

```
## IF THIS RETURNS TRUE THEN JUST STOP NOW THE DATA EXISTS:
User.where(email:'local_disco_user@artsymail.com').count > 0

app = ClientApplication.where(name: 'Metaphysics').first
user = User.create!(email: 'local_disco_user@artsymail.com', password: 'localdisco_is_great')
token = UserTrust.create_for_token_authentication(app, user: user, expires_in: 10.years.from_now)

artist1 = Artist.create!(name: 'Cloud Strife')
follow_artist_1 = FollowArtist.create!(user: user, artist: artist1)
art1 = Artwork.create!(artist_ids: [artist1.id], title: 'Still Life with Materia', private: false, published: true)
art2 = Artwork.create!(artist_ids: [artist1.id], title: 'Sephiroth Portrait', private: false, published: true)

artist2 = Artist.create!(name: 'Squall Leonheart')
art3 = Artwork.create!(artist_ids: [artist2.id], title: 'Laguna Again?!', private: false, published: true)
art4 = Artwork.create!(artist_ids: [artist2.id], title: 'Boss Battle', private: false, published: true)

partner1 = Partner.create!(given_name: 'not_gallery_followed')
follow_partner_1 = FollowProfile.create!(user: user, profile: partner1.profile)
show1 = PartnerShow.create!(partner: partner1, start_at: 10.years.ago, end_at: 10.years.from_now)
show1.add_artwork! art1

partner2 = Partner.create!(given_name: 'not_gallery_not_followed')  
show2 = PartnerShow.create!(partner: partner2, start_at: 10.years.ago, end_at: 10.years.from_now)
show2.add_artwork! art2

gallery1 = PartnerGallery.create!(given_name: 'gallery_followed')
follow_gallery_1 = FollowProfile.create!(user: user, profile: gallery1.profile)
show3 = PartnerShow.create!(partner: gallery1, start_at: 10.years.ago, end_at: 10.years.from_now)
show3.add_artwork! art3

gallery2 = PartnerGallery.create!(given_name: 'gallery_not_followed')
show4 = PartnerShow.create!(partner: gallery2, start_at: 10.years.ago, end_at: 10.years.from_now)
show4.add_artwork! art4

fair = Fair.create!(name: 'local_disco_fair', start_at: 15.years.ago, end_at: 15.years.from_now, published: true)
fair.partner_shows << show1
fair.partner_shows << show2
fair.partner_shows << show3
fair.partner_shows << show4

FairPartnerShowIndex.reindex_fair fair
FairService.invalidate_shows_and_artworks_caches fair
puts "Fair ID: #{fair.id}"
puts "X-Access-Token: #{token}"
```

Here's a graphQL query to test this functionality (the fair ID was generated by the above script when I ran it):

```
query {
  fair(id:"5c522f0cbc007d0007b4761b") {
    id
    followed_content{
      artists {
        name
      }
      galleries {
        name
      }
    }
  }
}
```

The response to this query should embed 'Cloud Strife' and 'gallery_followed' as the artist and gallery name, respectively, from the dataset generated. If you get Squall back as an artist too then this query is ignoring your follows. If you get `not_gallery_followed` then this query is ignoring the constraint to only return `PartnerGallery` types and leave out `Partner` types. If you get `gallery_not_followed` it's like getting Squall back, it's part of the fair but not something you personally follow so this query should not return it.

### CODE REVIEW

File: src/schema/fair.ts [hunk 1]
```diff
@@ -26,6 +28,19 @@ import ShowSort from "./sorts/show_sort"
 import { allViaLoader } from "lib/all"
 import { FairArtistSortsType } from "./sorts/fairArtistSorts"
 
+const FollowedContentType = () =>
```
Comment (mzikherman) [inline src/schema/fair.ts]: Curious, why the defer here? If the issue is referring to the `Artist` or `Partner` types and circular/diamond dependencies, you can do move that to just the `fields` part:  `fields: () => { ...`.

Comment (mbilokonsky) [reply src/schema/fair.ts]: Ah, that was exactly the issue, but your solution is cleaner, thanks! :)

File: src/schema/fair.ts [hunk 2]
```diff
@@ -26,6 +28,18 @@ import ShowSort from "./sorts/show_sort"
 import { allViaLoader } from "lib/all"
 import { FairArtistSortsType } from "./sorts/fairArtistSorts"
 
+const FollowedContentType = new GraphQLObjectType({
+  name: "followed_fair_content",
```
Comment (anandaroop) [inline src/schema/fair.ts]: I'd think this name would be simply `FollowedContent`, following convention?

**Comment by mbilokonsky** [comment]
@anandaroop The app and token are just there to generate an X-Access-Token for the user account we're following this with.

**Comment by mbilokonsky** [comment]
Ok, finally got this figured out. I was really hung up on fair and artist caches but it was the garner user cache that was throwing everything off - @sweir27 helped me figure that out, then we cleared it, now we have data!

So, this query: 

```query {
  fair(id:"5c522f0cbc007d0007b4761b") {
    id
    followed_content{
      artists {
        name
        _id
      }
      galleries {
        name
        _id
      }
    }
  }
}```

returns this result: 
```
{
  "data": {
    "fair": {
      "id": "local-disco-fair",
      "followed_content": {
        "artists": [
          {
            "name": "Strife Cloud",
            "_id": "5c522bd7bc007d0007b475ea"
          }
        ],
        "galleries": [
          {
            "name": "gallery_followed",
            "_id": "5c522effbc007d0007b47612"
          }
        ]
      }
    }
  },
  "extensions": {
    "requests": {
      "gravity": {
        "requests": {
          "fair/5c522f0cbc007d0007b4761b": {
            "time": "0s 4.398ms",
            "cache": true
          },
          "/me/follow/profiles?fair_id=local-disco-fair&owner_types%5B%5D=PartnerGallery": {
            "time": "0s 144.665ms",
            "cache": false
          },
          "me/follow/artists?fair_id=local-disco-fair": {
            "time": "0s 148.415ms",
            "cache": false
          }
        }
      }
    },
    "requestID": "9712b380-2599-11e9-a0ba-45a0a551c194"
  }
}
```

I think this is ready for formal review and merging now! Thanks all!

File: src/schema/fair.ts [hunk 3]
```diff
@@ -26,6 +28,18 @@ import ShowSort from "./sorts/show_sort"
 import { allViaLoader } from "lib/all"
 import { FairArtistSortsType } from "./sorts/fairArtistSorts"
 
+const FollowedContentType = new GraphQLObjectType({
+  name: "followed_content",
```
Comment (orta) [inline src/schema/fair.ts]: Minor: The type name should be `FollowedContent` to be consistent with everything else (you can see in the SDL how it looks a bit weird `followed_content: followed_content`)

File: src/schema/fair.ts [hunk 4]
```diff
@@ -56,6 +70,29 @@ const FairType = new GraphQLObjectType({
...
+          artists: followedArtistsLoader({ fair_id }).then(({ body }) => {
+            return body.map(artist_follow => artist_follow.artist)
+          }),
+          galleries: followedPartnersLoader({
+            fair_id,
+            owner_types: ["PartnerGallery"],
+          }).then(({ body }) => {
+            return body.map(profile_follow => profile_follow.profile.owner)
+          }),
+        }
```
Comment (orta) [inline src/schema/fair.ts]: is it worth thinking about the cases when you may only want artists or galleries, or do we always used both?

Comment (mbilokonsky) [reply src/schema/fair.ts]: As far as I know this is a very specific use case that requires both, but I see your point - ought this be further decomposed?

**Comment by mbilokonsky** [comment]
Hey @l2succes after speaking with @ashfurrow it sounds like there's no need to add shows here. This can be merged as-is, or if you'd like I can pick up @orta's point and generalize this to allow individual queries, if we think we'll need that?

**Comment by l2succes** [comment]
@mbilokonsky Let's go ahead and merge it as is. I can't think of a use case as of now where we'd be requesting one without the other.

**Pull request merged**